### ISSUE TITLE
Improve `prompt-buffer`'s suggestions narrowing

### ISSUE THREAD

**Issue by aadcg** [atlas-engineer/nyxt#2631]
After #2579, we now have a much richer `describe` interface!

But it seems to struggle narrowing down when matching user input with suggestions.

For instance, issue `describe-any` and input `"prompter:return-actions"`. The user needs to issue `select-next-source` until the matching  source is selected.

CC @Ambrevar

**Comment by aartaka** [comment]
There are several problems that you hint at:
- Not all sources are agressive enough in deleting irrelevant suggestions.
- `describe-*` source having no heuristics for, e.g., package prefixes.
- Fuzzy matching algorithm unpredictability. While it works for the tests and simple use-cases, I often find myself lost in suggestions :( Here we maybe need some algo that's more typo-aware or otherwise intuitive. [Maybe from here (we used to use Levenstein distance)](https://itnext.io/string-similarity-the-basic-know-your-algorithms-guide-3de3d7346227).

**Comment by Ambrevar** [comment]
One thing we do with remembrance mode is have 2 sources: one for exact matches, and one for non-exact matches.

It's very convenient, but in the case of describe-any it would make for many, many sources.

**Comment by aadcg** [comment]
> Fuzzy matching algorithm unpredictability. While it works for the tests and simple use-cases, I often find myself lost in suggestions :( Here we maybe need some algo that's more typo-aware or otherwise intuitive. Maybe from here (we used to use Levenstein distance).

Instead of jumping right into more advanced algorithms there's a simpler way to approach it.  If you define several completion algorithms (even if they're dead basic) and establish an order on their composition, you'll get a powerful completion system.

**Comment by Ambrevar** [comment]
> establish an order on their composition,

Precisely this is hard to figure out and might yield suboptimal results in practice :/

**Comment by aadcg** [comment]
> > establish an order on their composition,
> 
> Precisely this is hard to figure out and might yield suboptimal results in practice :/

Can you expand?

**Comment by Ambrevar** [comment]
I mean that figuring out an order between N sources that returns the most relevant suggestion _ordering_ is difficult.

In practice, some input might give a relevant ordering, while some other input might not.

**Comment by aadcg** [comment]
We're talking about different things, as I feared.

The problem you're referring to is: when matching INPUT against N lists of suggestions (N sources), it's easy to order the suggestions within each source, but hard to understand how sources should be ordered. Hard because we can define source ranking/ordering in multiple ways and, a priori, there's no way of knowing which heuristic is better.

I'm talking about a much simpler problem: matching INPUT against a list of suggestions (a single source). And I'm arguing that our `filter` is not implemented in the most general case. Instead of accepting a single function (or a symbol) it should accept a list of them. The order of the functions on that determines the order of composition. When a function return nil matches, the next one is used. Example: Set that list to be `(list #'prefix #'regex)`. Say you want to match "pot" against "potato". Since `prefix` will return matches, `regex` will never be called. When you match "po.*o", then `prefix` will return nil matches, and so `regex` kicks in.

I remind that I've raised this topic since @aartaka suggested implementing more complex algorithms. I counter-argued that supporting a handful of really basic algorithms and composing them makes up for a powerful filtering (completion) system. `prefix`, `substring`, `regex` and `fuzzy` would be quite powerful already. 

I could list other basic matching/filtering/completion styles, namely involving a special treatment of whitespaces, but I've illustrated my point.

Remark. Notice, however, that implementing `substring` requires having access to the location of the cursor. Example "foo|bar", where "|" denotes the location of the cursor completes according to the following regex `.*foo.*bar.*`. "f|oobar" would complete `.*f.*oobar.*`.

**Comment by aartaka** [comment]
> I remind that I've raised this topic since @aartaka suggested implementing more complex algorithms. I counter-argued that supporting a handful of really basic algorithms and composing them makes up for a powerful filtering (completion) system. `prefix`, `substring`, `regex` and `fuzzy` would be quite powerful already.
>
> I could list other basic matching/filtering/completion styles, namely involving a special treatment of whitespaces, but I've illustrated my point.

If you want that, you can always configure/redefine `prompter:source` to have a different matching function, no need to change anything anything beyond that architecturally.

But then, why I'm suggesting to have a better algorithm is that "defaults matter™️". If we have a good string search algo, we don't have to compose anything except in extremely rare cases. 

So my suggestion is:
- Make a default algo better.
- Preserve the configurability of the prompter entities (maybe make `promtper-source` a `user-class` or inherit `nyxt:source` from `prompter:source`?)
- And reconfigure matching algo for special cases like Lisp symbols search and whatnot.

> Remark. Notice, however, that implementing `substring` requires having access to the location of the cursor. Example "foo|bar", where "|" denotes the location of the cursor completes according to the following regex `.*foo.*bar.*`. "f|oobar" would complete `.*f.*oobar.*`.

Right, this one wouldn't hurt.

**Comment by aadcg** [comment]
> If you want that, you can always configure/redefine prompter:source to have a different matching function, no need to change anything anything beyond that architecturally.

That is not correct. I can set `filter` to whatever function I want, but I can't pass a list of them.

**Comment by aadcg** [comment]
> If we have a good string search algo, we don't have to compose anything except in extremely rare cases.

> So my suggestion is: Make a default algo better.

You can't say that any heuristic reasoning is better than the other. All of these algorithms are heuristic. I find our default, `fuzzy` to be very good actually. What is lacking is precisely composition: I like to order my completion mechanisms from strict (say `prefix`) to lenient  (`fuzzy` of `regex`).

**Comment by aadcg** [comment]
Regardless, @Ambrevar has raised the most urgent issue:

> I mean that figuring out an order between N sources that returns the most relevant suggestion ordering is difficult.

**Comment by Ambrevar** [comment]
> > 
> 
> That is not correct. I can set `filter` to whatever function I want, but I can't pass a list of them.

You misunderstood artyom: a list of functions is not a good idiom, since functions can be composed in arbitrary ways (functional programming), which is strictly more powerful than a "list of functions".

In other words, you can set the filter to `(alex:conjoin fun1 fun2 ...)`, a currified list of functions, etc.  Thus no need to support list of functions, functional programming gives it for free.

This is how the prompter library was designed, with composition in mind.

**Comment by aadcg** [comment]
I see it now! That's a better idea indeed.
